### PR TITLE
[CO-349] Implement throttling middleware

### PR DESCRIPTION
**Pull request by parsonsmatt**
## Description

This PR implements the `wai-middleware-throttle` in the new wallet code. Configuration via the config file will be implemented in a follow-up PR.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CO-349

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: pkgs/default.nix [hunk 1]
```diff
@@ -17801,6 +17801,7 @@ license = stdenv.lib.licenses.mit;
 , vector
 , wai
 , wai-cors
+, wai-middleware-throttle
```
Comment (parsonsmatt) [inline pkgs/default.nix]: This library is [small](https://github.com/creichert/wai-middleware-throttle/blob/master/Network/Wai/Middleware/Throttle.hs), [has a test-suite](https://github.com/creichert/wai-middleware-throttle/blob/master/test/WaiMiddlewareThrottleSpec.hs), is relatively [well-documented](http://hackage.haskell.org/package/wai-middleware-throttle-0.2.2.1/docs/Network-Wai-Middleware-Throttle.html), and is [actively maintained](https://github.com/creichert/wai-middleware-throttle/commits/master).

The [only alternatve on Hackage](http://hackage.haskell.org/package/wai-throttler) has none of these properties.

File: wallet-new/src/Cardano/Wallet/Server/Plugins.hs [hunk 2]
```diff
@@ -123,31 +123,33 @@ legacyWalletBackend :: (HasConfigurations, HasCompileInfo)
...
+                usingLoggerName "NodeIPC"
+                $ flip runReaderT ctx
+                $ startNodeJsIPC port
+        walletServeImpl
+            (getApplication diffusion)
+            walletAddress
+            -- Disable TLS if in debug mode.
+            (if isDebugMode walletRunMode then Nothing else walletTLSParams)
+            (Just $ setOnExceptionResponse exceptionHandler defaultSettings)
+            (Just portCallback)
```
Comment (parsonsmatt) [inline wallet-new/src/Cardano/Wallet/Server/Plugins.hs]: spacing change to bring code into compliance with 4 space indentation

File: wallet-new/src/Cardano/Wallet/Server/Plugins.hs [hunk 3]
```diff
@@ -204,23 +206,23 @@ walletBackend protocolMagic (NewWalletBackendParams WalletBackendParams{..}) (pa
...
+            let
+                portCallback :: Word16 -> IO ()
+                portCallback port = usingLoggerName "NodeIPC" $ flip runReaderT ctx $ startNodeJsIPC port
+            walletServeImpl
+                (getApplication active)
+                walletAddress
+                -- Disable TLS if in debug modeit .
+                (if isDebugMode walletRunMode then Nothing else walletTLSParams)
+                Nothing
+                (Just portCallback)
```
Comment (parsonsmatt) [inline wallet-new/src/Cardano/Wallet/Server/Plugins.hs]: spacing change to bring code into compliance with 4 space indentation

File: wallet-new/src/Cardano/Wallet/Server/Plugins.hs [hunk 4]
```diff
@@ -123,31 +123,33 @@ legacyWalletBackend :: (HasConfigurations, HasCompileInfo)
...
   where
     -- Gets the Wai `Application` to run.
     getApplication :: Diffusion WalletWebMode -> WalletWebMode Application
     getApplication diffusion = do
         logInfo "Wallet Web API has STARTED!"
         wsConn <- getWalletWebSockets
         ctx <- V0.walletWebModeContext
-        return
-            $ withMiddleware walletRunMode
+        withMiddleware walletRunMode
```
Comment (parsonsmatt) [inline wallet-new/src/Cardano/Wallet/Server/Plugins.hs]: `withMiddleware` is now a monadic function, so we can replace `return . withMiddleware` with `withMiddleware` now.

File: wallet-new/src/Cardano/Wallet/Server/Plugins.hs [hunk 5]
```diff
@@ -246,10 +248,14 @@ syncWalletWorker = pure $ const $
...
-withMiddleware wrm app
-  | isDebugMode wrm = corsMiddleware app
-  | otherwise = app
+withMiddleware :: MonadIO m => RunMode -> Application -> m Application
+withMiddleware wrm app = do
+    st <- liftIO $ Throttle.initThrottler
+    pure
+        . (if isDebugMode wrm then corsMiddleware else identity)
+        . Throttle.throttle Throttle.defaultThrottleSettings st
+        $ app
```
Comment (parsonsmatt) [inline wallet-new/src/Cardano/Wallet/Server/Plugins.hs]: This is the main change to add throttling. A future PR will bring customized throttling settings from the `configuration.yaml` file.

File: wallet-new/src/Cardano/Wallet/Server/Plugins.hs (hunk 3 above)
Comment (KtorZ) [reply wallet-new/src/Cardano/Wallet/Server/Plugins.hs]: Side-note but, better make all those "linting" changes or import re-organizing in a separated commit with a clear message. It helps _A LOT_ during rebases / merge-conflicts :)

File: wallet-new/src/Cardano/Wallet/Server/Plugins.hs (hunk 5 above)
Comment (KtorZ) [reply wallet-new/src/Cardano/Wallet/Server/Plugins.hs]: I am curious to see whether this is going to break things like my addition of the HTTP/2.0 middleware breaks things :thinking: ... There's something weird 'about WAI middlewares.

**Comment by KtorZ** [comment]
Hydra says it doesn't like your code:

```
/nix/store/hdbr6240s17l7slgljd5n0ylpl0kjb4s-clang-wrapper-5.0.1/bin/cc: line 183: /nix/store/2vk76kbkb0n1fqwwnppml0qb98j13h1k-clang-5.0.1/bin/clang: Argument list too long
```

**Comment by KtorZ** [comment]
```
wal-integr-test: FailureResponse (Response {responseStatusCode = Status {statusCode = 429, statusMessage = "Too Many Requests"}, responseHeaders = fromList [("Transfer-Encoding","chunked"),("Date","Mon, 06 Aug 2018 22:41:48 GMT"),("Server","Warp/3.2.22"),("Content-Type","application/json")], responseHttpVersion = HTTP/1.1, responseBody = "{\"message\":\"Too many requests.\"}"})
```

Hahaha, at least this answers my previous question :joy: ... Any ideas about what are the default parameters for the throttling? Perhaps would be good to make that configurable as part of this story already...

**Comment by parsonsmatt** [comment]
@KtorZ The default settings are [listed here, in source code](http://hackage.haskell.org/package/wai-middleware-throttle-0.2.2.1/docs/src/Network.Wai.Middleware.Throttle.html#defaultThrottleSettings).

I'll modify the response to be JSend :)

File: wallet-new/src/Cardano/Wallet/Server/Plugins.hs (hunk 3 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/Server/Plugins.hs]: hmm! I tend to do squash-n-merge on GitHub when I am done with a PR, which obviously rmeoves that distinction. Should I just do merge commits instead?

**Comment by parsonsmatt** [comment]
As far as testing this: We can duplicate [the testing that the library has done](https://github.com/creichert/wai-middleware-throttle/blob/master/test/WaiMiddlewareThrottleSpec.hs), but I'm not sure how useful that is?

**Comment by KtorZ** [comment]
```hs
defaultThrottleSettings
    = ThrottleSettings {
        isThrottled         = return . const True
      , throttleRate        = 1 :: Integer  -- req / throttlePeriod
      , throttlePeriod      = 1000000 :: Integer -- microseconds
      , throttleBurst       = 1 :: Integer  -- concurrent requests
      , onThrottled         = onThrottled'
      , onRequestError      = onRequestError'
      }
```

Wow.. One request per second at max, that's harsh ! We definitely need to make this configurable, and allow throttling to be disabled from the configuration as well.

What do you think of having it  disabled by default?

**Comment by parsonsmatt** [comment]
Well, the ability to configure it and load that configuration from a file is going to come in the next PR :) I suppose we can fairly easily have the default be either way.

**Comment by KtorZ** [comment]
Okay, then let's go for hard-coded default for now, without what the CI won't ever pass haha. We make at least 10 req/s if not more already in some cases :p

**Comment by parsonsmatt** [comment]
The integration tests are failing on the throttling still. I'm going to merge as-is now and will ensure that in testing there are no throttling concerns.

**Pull request merged**